### PR TITLE
[#10245] UAT Fix: Instructor Search Page

### PR DESCRIPTION
**Pull request by madanalogy**
Part of #10245 

> Instructor search page: If an instructor who is admin searches for students, s/he will get all the matching students in the system. This is not accurate as the intention of this page is not for system-wide search.

Added new param `entitytype` for student search

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java [hunk 1]
```diff
@@ -72,11 +79,12 @@ private String findAvailableInstructorGoogleIdForCourse(String courseId) {
     @Override
     public ActionResult execute() {
         String searchKey = getNonNullRequestParamValue(Const.ParamsNames.SEARCH_KEY);
+        String entity = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
         List<StudentAttributes> students;
         List<StudentData> studentDataList = new ArrayList<>();
 
         // Search for students
-        if (userInfo.isAdmin) {
+        if (userInfo.isAdmin && "admin".equals(entity)) {
```
Comment (t-cheepeng) [inline src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java:87]: same here

File: src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java [hunk 2]
```diff
@@ -86,7 +94,7 @@ public ActionResult execute() {
         for (StudentAttributes s : students) {
             StudentData studentData = new StudentData(s);
 
-            if (userInfo.isAdmin) {
+            if (userInfo.isAdmin && "admin".equals(entity)) {
```
Comment (t-cheepeng) [inline src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java:97]: same here

File: src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java [hunk 3]
```diff
@@ -28,6 +28,13 @@ public void checkSpecificAccessControl() {
         if (!userInfo.isInstructor && !userInfo.isAdmin) {
             throw new UnauthorizedAccessException("Instructor or Admin privilege is required to access this resource.");
         }
+        String entity = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+        if (!userInfo.isAdmin && "admin".equals(entity)) {
```
Comment (t-cheepeng) [inline src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java:32]: Maybe extract the `admin` and 'instructor' string to a variable/const if possible?

```suggestion
        if (!userInfo.isAdmin && entity.equals("admin")) {
```

Comment (madanalogy) [reply src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java:32]: Yeap sounds like a good idea, I’ll extract to a constant. The comparison order you suggested is not allowed by our lint settings, something about comparison of literals having to come first haha

**Comment by wkurniawan07** [comment]
@madanalogy have you checked the front-end? Changing the response code from 4XX (failure) to 2XX (success) means the code execution flow will change.

**Comment by madanalogy** [comment]
> @madanalogy have you checked the front-end? Changing the response code from 4XX (failure) to 2XX (success) means the code execution flow will change.

I've manually navigated through the student profile pages and submission pages, everything seems to be okay

**Comment by madanalogy** [comment]
After a lot of research and many long hours spent trying to refactor `404` responses into `204` responses, I believe that we shouldn't because of two primary reasons: the first being that I am unable to find a feasible implementation and the second being that even if there was one, `404` is actually much more appropriate in where it's currently being used.

**On feasible implementations:**

The instances already implementing 404 intentionally (e.g. feedback response comments and student profile picture) do so because Angular is able to catch 404 responses as an error thus allowing for separate response handlers for management of program flow. 204 does not count as a http error and thus is indistinguishable from a successful server response.

One alternative I tried was to check for the state of the observable, whether it was an instance of the required output type. This is theoretically possible because upon returning a 404 the backend sends a `MessageOutput` instead of the expected data output type. This however has to be checked and transformed in the service itself, which is not impossible but requires that calling components perform null checks on the received data.

Another alternative I tried was to have the backend send the intended data output type but as a null. This was not possible for feedback response comments as it did not have a null constructor. This also however runs into the same problem of requiring additional code for null checks in the frontend calling components.

**On inappropriate use of status code:**

The 404 status code is supposed to indicate that the requested for resource is not found, which from the perspective of the backend makes sense. Attempting to request for a comment that does not exist should carry a response that the comment does not exist. Source: [404 Not Found](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404)

The 204 specification RFC however states that "The 204 (No Content) status code indicates that the server has successfully fulfilled the request and that there is no additional content to send in the response payload body." This is not applicable in our use case because the backend would still respond with a payload body (the `MessageOutput`) indicating that the requested for resource is not found. Sources: [204 No Content](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204), [RFC 7231](https://tools.ietf.org/html/rfc7231#section-6.3.5)

**Moving forward:**
I've added `404` response management to the student profile calls to be consistent with similar implementations in the frontend. Moving forward with this issue I believe would take up too much development time to be worth the cost during this period considering that the impact to the end user isn't immediately noticeable. Might be worthwhile opening a separate issue tracker to carry forward this discussion.

**Comment by wkurniawan07** [comment]
@madanalogy I suggest to exclude the status code-related changes from this PR

File: src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java [hunk 4]
```diff
@@ -28,6 +28,13 @@ public void checkSpecificAccessControl() {
         if (!userInfo.isInstructor && !userInfo.isAdmin) {
             throw new UnauthorizedAccessException("Instructor or Admin privilege is required to access this resource.");
         }
+        String entity = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java:31]: If entity type equals to something else, we might have unexpected flow in both `execute()` and `checkSpecificAccessControl()`

Comment (madanalogy) [reply src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java:32]: Added throw exception if entities do not match exactly

Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/SearchStudentsAction.java:32]: Don't forget test case :P

File: src/test/java/teammates/test/cases/webapi/SearchStudentsActionTest.java [hunk 5]
```diff
@@ -37,9 +37,38 @@ protected void testExecute() {
...
+
+        String[] invalidEntityParams = new String[] {
+                Const.ParamsNames.SEARCH_KEY, "dummy",
+                Const.ParamsNames.ENTITY_TYPE, "dummy",
+        };
+        verifyHttpParameterFailure(invalidEntityParams);
+    }
+
+    @Test
+    public void execute_wrongEntityType_parameterFailure() {
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/webapi/SearchStudentsActionTest.java:57]: This needs not be a separate test case as `wrongEntityType` is a subset of `invalidParameters`

**Pull request merged**